### PR TITLE
feat(eslint-plugin): [block-spacing] extending base rule for TS related blocks

### LINKED ISSUE #5990: Base rule extension: block-spacing Extending base rule for TS related blocks
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+base+rule+extension%22) and found none that match my proposal.
- [X] I have searched the [current extension rule list](https://typescript-eslint.io/rules/#extension-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the base rule

https://eslint.org/docs/latest/rules/block-spacing

### Description

I believe the purpose of the base rule, applies for TS too

### Fail

```typescript
interface Shape {x: number}
interface Shape { x: number}
interface Shape {x: number }
interface Shape {x: number; y: number}
```

### Pass

```typescript
interface Shape { x: number }
interface Shape { x: number; y: number }
```

### Additional Info

Actually, having a rule that rejects single line `interface` declarations could work too

### PR DESCRIPTION
**Pull request by omril1**
## PR Checklist

- [x] Addresses an existing open issue: fixes #5990
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

### CODE REVIEW

File: packages/eslint-plugin/src/rules/block-spacing.ts
Comment (omril1) [inline packages/eslint-plugin/src/rules/block-spacing.ts:46]: The only significant change from the base rule, it worked flawlessly for interface and object type literal.
Added a filter for punctuator because enums don't have a body node in the AST

https://github.com/eslint/eslint/blob/3ae0574fc78c4a2b406625e4792cb2859cb9bcb1/lib/rules/block-spacing.js#L60

File: packages/eslint-plugin/src/rules/block-spacing.ts
Comment (omril1) [inline packages/eslint-plugin/src/rules/block-spacing.ts:163]: Kept the base rule for interface and type literal, though changing to `checkSpacingInsideBraces` will work as well

File: packages/eslint-plugin/src/rules/block-spacing.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/block-spacing.ts:70]: Nit: we have types in TypeScript. `ASTNode` is imprecise and the `@returns` is implied.

```suggestion
     * @param node A BlockStatement/StaticBlock/SwitchStatement node to check.
```

Seems like these were copy & pasted from https://github.com/eslint/eslint/blob/ba74253e8bd63e9e163bbee0540031be77e39253/lib/rules/block-spacing.js#L86? I don't see other base rule files giving explicit attribution, so I guess this should be fine as-is.

File: packages/eslint-plugin/src/rules/block-spacing.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/block-spacing.ts:97]: Indeed. If this isn't covered by tests, then either:

* We need more tests
* This is unreachable and can be removed.

File: packages/eslint-plugin/src/rules/block-spacing.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/block-spacing.ts:99]: Nit: the comment is a little redundant, I think.

```suggestion
```

File: packages/eslint-plugin/tests/rules/block-spacing.test.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/block-spacing.test.ts:17]: Nit: other arrays are named as the plural. Let's keep it consistent for this one.

```suggestion
const typeDeclarations = [
```

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Mostly looks great to me! Just some touchups around untested code, and a few nitpicks. Thanks for sending this in & working on it! ✨

File: packages/eslint-plugin/src/rules/block-spacing.ts
Comment (omril1) [reply packages/eslint-plugin/src/rules/block-spacing.ts:70]: I saw other rules that copied from the base rule, so I assumed it's ok, do you suggest only removing the types from the jsdoc?

Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/block-spacing.ts:70]: Yeah there's no need for the jsdoc.

**Comment by omril1** [comment]
Sorry for the force push, something in vscode messed up my commits again

File: packages/eslint-plugin/src/rules/block-spacing.ts
Comment (omril1) [reply packages/eslint-plugin/src/rules/block-spacing.ts:95]: Added more tests, and now this is covered

**Review by JoshuaKGoldberg — DISMISSED** [review]
Looks great to me, thanks for sending this in!

Because this is an extension of a base rule, I'd feel more comfortable getting a review from another maintainer. Just to be safe 😄 .

File: packages/eslint-plugin/src/rules/block-spacing.ts
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/block-spacing.ts]: Ok just following up from internal discussions! Still no need to take action here. Just posting as a reference...

Some of our extension rules include code that's very similar to the code from the base ESLint rule. Sometimes there are entire lines or even functions that look the same.

[ESLint is licensed under the MIT License](https://github.com/eslint/eslint/blob/b401cde47d44746ff91b8feced3fb3a4e32c0e12/LICENSE). MIT stipulates _(emphasis mine)_:

> Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
> 
> The above copyright notice and this permission notice shall be included in
all **copies or substantial portions of the Software**.

We're not legal experts -let alone copyright lawyers- but we believe we're not in violation of this because:

* Our code does not copy substantial portions. At most, the _ideas_ of the original code are allowed in as reference, but we use our own code style & standards in the implementation.
* We link to the original ESLint source quite often in our site & packages, including in the individual docs pages for rules.

_If anybody knows my statements to be wrong and/or that we need to do something else, please let us know!_

**Review by bradzacher — APPROVED** [review]

**Pull request merged**
